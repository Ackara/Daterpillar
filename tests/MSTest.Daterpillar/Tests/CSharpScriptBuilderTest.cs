using Ackara.Daterpillar.Scripting;
using ApprovalTests;
using ApprovalTests.Reporters;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Shouldly;
using System;
using System.Collections.Generic;
using System.Linq;

namespace MSTest.Daterpillar.Tests
{
    [TestClass]
    [UseReporter(typeof(FileLauncherReporter), typeof(ClipboardReporter))]
    public class CSharpScriptBuilderTest
    {
        [TestMethod]
        public void Append_should_return_a_csharp_script_that_defines_a_class_when_all_settings_are_enabled()
        {
            // Arrange
            var schema = MockData.GetSchema(FName.scriptingTest_partial_schemaXML);
            var sut = new CSharpScriptBuilder(new CSharpScriptBuilderSettings()
            {
                AddConstants = true,
                IgnoreComments = false,
                AddSchemaAttributes = true,
                UseVirtualProperties = true,
                InheritEntityBaseClass = true,
                AddDataContractAttributes = true,
            });

            // Act
            sut.Append(schema);
            var script = sut.GetContent();
            var codeIsCompilable = TryCompilation(ref script, out string errorMsg);

            // Assert
            script.ShouldNotBeNullOrEmpty();
            Approvals.Verify(script);
            codeIsCompilable.ShouldBeTrue(errorMsg);
        }

        [TestMethod]
        public void Append_should_return_a_csharp_script_that_defines_a_class_when_all_settings_are_disabled()
        {
            // Arrange
            var schema = MockData.GetSchema(FName.scriptingTest_partial_schemaXML);
            var sut = new CSharpScriptBuilder(new CSharpScriptBuilderSettings()
            {
                AddConstants = false,
                IgnoreComments = true,
                AddSchemaAttributes = false,
                UseVirtualProperties = false,
                InheritEntityBaseClass = false,
                AddDataContractAttributes = false,
            });

            // Act
            sut.Append(schema);
            var script = sut.GetContent();
            var codeIsCompilable = TryCompilation(ref script, out string errorMsg);

            // Assert
            script.ShouldNotBeNullOrEmpty();
            Approvals.Verify(script);
            codeIsCompilable.ShouldBeTrue(errorMsg);
        }

        [TestMethod]
        public void Append_should_return_a_csharp_script_that_defines_a_enum()
        {
            // Arrange
            var sample = new Dictionary<string, int>()
            {
                {"Red", 0 },
                {"Green", 1 },
                {"Blue", 2 },
                {"Dark Gray", 4 },
            };

            var sut = new CSharpScriptBuilder(new CSharpScriptBuilderSettings()
            {
                IgnoreComments = false,
                AddSchemaAttributes = true,
                UseVirtualProperties = true,
                InheritEntityBaseClass = true,
            });

            // Act
            sut.Append("Color", sample);
            var code = sut.GetContent();
            var codeIsCompilable = TryCompilation(ref code, out string errorMsg);

            // Assert
            code.ShouldNotBeNullOrEmpty();
            Approvals.Verify(code);
            codeIsCompilable.ShouldBeTrue(errorMsg);
        }

        [TestMethod]
        public void Append_should_return_a_csharp_script_that_defines_a_enum_that_supports_flags()
        {
            // Arrange
            var sample = new Dictionary<string, int>()
            {
                {"Red", 0 },
                {"Green", 1 },
                {"Blue", 2 },
                {"Dark Gray", 4 },
            };

            var sut = new CSharpScriptBuilder(new CSharpScriptBuilderSettings()
            {
                IgnoreComments = false,
                AddSchemaAttributes = true,
                UseVirtualProperties = true,
                InheritEntityBaseClass = true,
            });

            // Act
            sut.Append("Color", sample, flag: true);
            var code = sut.GetContent();
            var codeIsCompilable = TryCompilation(ref code, out string errorMsg);

            // Assert
            code.ShouldNotBeNullOrEmpty();
            Approvals.Verify(code);
            codeIsCompilable.ShouldBeTrue(errorMsg);
        }

        private static bool TryCompilation(ref string code, out string errorMsg)
        {
            var errors = CSharpSyntaxTree.ParseText(code).GetDiagnostics();
            errorMsg = string.Join(Environment.NewLine, errors.Select(x => $"[{x.Severity}] {x.GetMessage()}"));

            if (errors.Count() > 0)
            {
                code = string.Format("/*{0}ERRORS!!!{0}{1}{0}*/{0}{0}{2}", Environment.NewLine, errorMsg, code);
            }

            return errors.Count() == 0;
        }
    }
}