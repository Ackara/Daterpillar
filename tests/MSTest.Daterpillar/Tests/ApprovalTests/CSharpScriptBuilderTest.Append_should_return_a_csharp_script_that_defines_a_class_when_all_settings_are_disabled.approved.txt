public partial class CardType
{
	public int Id { get; set; }

	public string Name { get; set; }

}

public partial class Attribute
{
	public int Id { get; set; }

	public string Name { get; set; }

}

public partial class CardIcon
{
	public int Id { get; set; }

	public string Name { get; set; }

}

public partial class MonsterType
{
	public int Id { get; set; }

	public string Name { get; set; }

}

public partial class Ability
{
	public int Id { get; set; }

	public string Name { get; set; }

}

public partial class Card
{
	public int Id { get; set; }

	public string Name { get; set; }

	public string Text { get; set; }

	public string PendulumEffect { get; set; }

	public int Level { get; set; }

	public int Atk { get; set; }

	public int Def { get; set; }

	public int CardTypeId { get; set; }

	public int AttributeId { get; set; }

	public int CardIconId { get; set; }

	public int MonsterTypeId { get; set; }

	public int AbilityId { get; set; }

}

public partial class CardExtras
{
	public int CardId { get; set; }

	public string Rulings { get; set; }

	public string Tips { get; set; }

	public string Trivia { get; set; }

	public int Passcode { get; set; }

}

public partial class Pack
{
	public int Id { get; set; }

	public string Acronym { get; set; }

	public string Name { get; set; }

	public string Description { get; set; }

	public DateTime ReleaseDate { get; set; }

	public int Size { get; set; }

	public int KonamiId { get; set; }

}

public partial class Rarity
{
	public int Id { get; set; }

	public string Name { get; set; }

	public string Code { get; set; }

	public int Value { get; set; }

}

public partial class CardNumber
{
	public string PackAcronym { get; set; }

	public int Number { get; set; }

	public int RarityId { get; set; }

	public int ArtIndex { get; set; }

	public int CardId { get; set; }

	public int PackId { get; set; }

}

public partial class Effect
{
	public int Id { get; set; }

	public string Name { get; set; }

}

